File: packages/buyerSdkPy/razoragent_buyer_sdk/agentMandateBuilder.py
"""AP2 Mandate lifecycle builder, canonical hashing, and cryptographic chain verifier."""

import time
from typing import Any, Dict, List, Optional
import uuid

from .agentKeyManager import (
    AgentKeyManager,
    canonicalizeJson,
    computeSha256Digest,
)
from .constants import (
    defaultCurrency,
    defaultIntentValiditySeconds,
)
from .exceptions import (
    MandateHashMismatchError,
    MandateValidationError,
)
from .models import (
    AmendmentMandate,
    CartItemSchema,
    CartMandate,
    ExecutionMandate,
    IntentMandate,
    TaxBreakdownSchema,
)


def _generateNonce() -> str:
    """Generates random single-use cryptographic nonce."""
    return f"nonce_{uuid.uuid4().hex}"


def computeMandateHash(mandateModel: Any) -> str:
    """Computes SHA-256 hash of unsigned mandate payload via RFC 8785 JCS canonicalization."""
    if hasattr(mandateModel, "model_dump"):
        dataDict = mandateModel.model_dump()
    elif isinstance(mandateModel, dict):
        dataDict = dict(mandateModel)
    else:
        raise ValueError("mandateModel must be a Pydantic model or dict")

    signatureKeys = {"userSignature", "merchantSignature", "agentSignature"}
    unsignedDict = {k: v for k, v in dataDict.items() if k not in signatureKeys}
    canonicalBytes = canonicalizeJson(unsignedDict)
    return computeSha256Digest(canonicalBytes)


def createIntentMandate(
    mandateId: str,
    userKeyManager: AgentKeyManager,
    delegatedAgentDid: str,
    maxBudgetPaise: int,
    upiCircleDelegationToken: str,
    singleTransactionLimitPaise: int,
    authorizedCategories: Optional[List[str]] = None,
    validUntilTimestamp: Optional[int] = None,
    nonce: Optional[str] = None,
    timestamp: Optional[int] = None,
) -> IntentMandate:
    """Constructs and signs a new IntentMandate (M_I)."""
    if maxBudgetPaise <= 0 or singleTransactionLimitPaise <= 0:
        raise MandateValidationError("maxBudgetPaise and singleTransactionLimitPaise must be > 0")

    ts = timestamp if timestamp is not None else int(time.time())
    non = nonce or _generateNonce()
    cats = authorizedCategories if authorizedCategories is not None else []
    validUntil = validUntilTimestamp if validUntilTimestamp is not None else (ts + defaultIntentValiditySeconds)

    payload: Dict[str, Any] = {
        "authorizedCategories": cats, "currency": defaultCurrency, "delegatedAgentDid": delegatedAgentDid,
        "mandateId": mandateId, "maxBudgetPaise": maxBudgetPaise, "nonce": non,
        "singleTransactionLimitPaise": singleTransactionLimitPaise, "timestamp": ts,
        "upiCircleDelegationToken": upiCircleDelegationToken, "userDid": userKeyManager.getAgentDid(),
        "validUntilTimestamp": validUntil,
    }

    sig = userKeyManager.signPayload(payload)
    return IntentMandate(
        mandateId=mandateId, userDid=userKeyManager.getAgentDid(), delegatedAgentDid=delegatedAgentDid,
        maxBudgetPaise=maxBudgetPaise, currency=defaultCurrency, authorizedCategories=cats,
        validUntilTimestamp=validUntil, upiCircleDelegationToken=upiCircleDelegationToken,
        singleTransactionLimitPaise=singleTransactionLimitPaise, nonce=non, timestamp=ts, userSignature=sig,
    )


def _buildCartMandatePayload(
    cartId: str, merchantDid: str, merchantGstin: str, merchantStateCode: str,
    buyerDeliveryPincode: str, buyerDeliveryStateCode: str, items: List[CartItemSchema],
    taxableSubtotalPaise: int, taxBreakdown: TaxBreakdownSchema, shippingPaise: int,
    discountPaise: int, totalPaise: int, inventoryLockToken: str,
    inventoryLockExpiresAt: int, nonce: str, timestamp: int,
) -> Dict[str, Any]:
    return {
        "buyerDeliveryPincode": buyerDeliveryPincode, "buyerDeliveryStateCode": buyerDeliveryStateCode,
        "cartId": cartId, "discountPaise": discountPaise, "inventoryLockExpiresAt": inventoryLockExpiresAt,
        "inventoryLockToken": inventoryLockToken, "items": [item.model_dump() for item in items],
        "merchantDid": merchantDid, "merchantGstin": merchantGstin,
        "merchantStateCode": merchantStateCode, "nonce": nonce, "shippingPaise": shippingPaise,
        "taxBreakdown": taxBreakdown.model_dump(), "taxableSubtotalPaise": taxableSubtotalPaise,
        "timestamp": timestamp, "totalPaise": totalPaise,
    }


def createCartMandate(
    cartId: str, merchantKeyManager: AgentKeyManager, merchantGstin: str,
    merchantStateCode: str, buyerDeliveryPincode: str, buyerDeliveryStateCode: str,
    items: List[CartItemSchema], taxableSubtotalPaise: int, taxBreakdown: TaxBreakdownSchema,
    shippingPaise: int, discountPaise: int, totalPaise: int,
    inventoryLockToken: str, inventoryLockExpiresAt: int,
    nonce: Optional[str] = None, timestamp: Optional[int] = None,
) -> CartMandate:
    """Constructs and signs a new CartMandate (M_C)."""
    ts = timestamp if timestamp is not None else int(time.time())
    non = nonce or _generateNonce()
    mDid = merchantKeyManager.getAgentDid()
    payload = _buildCartMandatePayload(
        cartId, mDid, merchantGstin, merchantStateCode, buyerDeliveryPincode,
        buyerDeliveryStateCode, items, taxableSubtotalPaise, taxBreakdown,
        shippingPaise, discountPaise, totalPaise, inventoryLockToken,
        inventoryLockExpiresAt, non, ts,
    )
    sig = merchantKeyManager.signPayload(payload)
    return CartMandate(
        cartId=cartId, merchantDid=mDid, merchantGstin=merchantGstin,
        merchantStateCode=merchantStateCode, buyerDeliveryPincode=buyerDeliveryPincode,
        buyerDeliveryStateCode=buyerDeliveryStateCode, items=items, taxableSubtotalPaise=taxableSubtotalPaise,
        taxBreakdown=taxBreakdown, shippingPaise=shippingPaise, discountPaise=discountPaise,
        totalPaise=totalPaise, inventoryLockToken=inventoryLockToken,
        inventoryLockExpiresAt=inventoryLockExpiresAt, nonce=non, timestamp=ts, merchantSignature=sig,
    )



def createExecutionMandate(
    executionId: str,
    buyerKeyManager: AgentKeyManager,
    intentMandate: IntentMandate,
    cartMandate: CartMandate,
    settlementAmountPaise: int,
    upiCircleToken: str,
    nonce: Optional[str] = None,
    timestamp: Optional[int] = None,
) -> ExecutionMandate:
    """Constructs and signs a new ExecutionMandate (M_E) binding Intent and Cart hashes."""
    ts = timestamp if timestamp is not None else int(time.time())
    non = nonce or _generateNonce()
    iHash = computeMandateHash(intentMandate)
    cHash = computeMandateHash(cartMandate)

    payload: Dict[str, Any] = {
        "buyerAgentDid": buyerKeyManager.getAgentDid(), "cartMandateHash": cHash,
        "currency": defaultCurrency, "executionId": executionId, "intentMandateHash": iHash,
        "nonce": non, "settlementAmountPaise": settlementAmountPaise, "timestamp": ts,
        "upiCircleToken": upiCircleToken,
    }

    sig = buyerKeyManager.signPayload(payload)
    return ExecutionMandate(
        executionId=executionId, buyerAgentDid=buyerKeyManager.getAgentDid(),
        intentMandateHash=iHash, cartMandateHash=cHash, settlementAmountPaise=settlementAmountPaise,
        currency=defaultCurrency, upiCircleToken=upiCircleToken, nonce=non, timestamp=ts, agentSignature=sig,
    )


def createAmendmentMandate(
    amendmentId: str,
    buyerKeyManager: AgentKeyManager,
    merchantKeyManager: AgentKeyManager,
    previousCartMandate: CartMandate,
    newCartMandate: CartMandate,
    substitutedSkuMapping: Dict[str, str],
    priceDeltaPaise: int,
    amendmentReason: str,
    nonce: Optional[str] = None,
    timestamp: Optional[int] = None,
) -> AmendmentMandate:
    """Constructs dual-signed AmendmentMandate (M_A) for out-of-stock or price updates."""
    ts = timestamp if timestamp is not None else int(time.time())
    non = nonce or _generateNonce()
    pHash = computeMandateHash(previousCartMandate)
    nHash = computeMandateHash(newCartMandate)

    payload: Dict[str, Any] = {
        "amendmentId": amendmentId, "amendmentReason": amendmentReason, "newCartMandateHash": nHash,
        "nonce": non, "previousCartMandateHash": pHash, "priceDeltaPaise": priceDeltaPaise,
        "substitutedSkuMapping": substitutedSkuMapping, "timestamp": ts,
    }

    cBytes = canonicalizeJson(payload)
    return AmendmentMandate(
        amendmentId=amendmentId, previousCartMandateHash=pHash, newCartMandateHash=nHash,
        substitutedSkuMapping=substitutedSkuMapping, priceDeltaPaise=priceDeltaPaise,
        amendmentReason=amendmentReason, nonce=non, timestamp=ts,
        agentSignature=buyerKeyManager.signCanonicalBytes(cBytes),
        merchantSignature=merchantKeyManager.signCanonicalBytes(cBytes),
    )


def verifyMandateHashChain(
    intentMandate: IntentMandate,
    cartMandate: CartMandate,
    executionMandate: ExecutionMandate,
    raiseOnMismatch: bool = True,
) -> bool:
    """Validates that ExecutionMandate binds exact cryptographic hashes of Intent and Cart."""
    expectedIntentHash = computeMandateHash(intentMandate)
    expectedCartHash = computeMandateHash(cartMandate)

    if executionMandate.intentMandateHash != expectedIntentHash:
        if raiseOnMismatch:
            raise MandateHashMismatchError(
                f"Intent hash mismatch: expected {expectedIntentHash}, got {executionMandate.intentMandateHash}"
            )
        return False

    if executionMandate.cartMandateHash != expectedCartHash:
        if raiseOnMismatch:
            raise MandateHashMismatchError(
                f"Cart hash mismatch: expected {expectedCartHash}, got {executionMandate.cartMandateHash}"
            )
        return False

    return True


def validateMandateInvariants(
    intentMandate: IntentMandate,
    cartMandate: CartMandate,
    executionMandate: ExecutionMandate,
    currentTime: Optional[int] = None,
) -> None:
    """Validates spending limits, temporal validity, and totals across the mandate chain."""
    now = currentTime if currentTime is not None else int(time.time())
    if now > intentMandate.validUntilTimestamp:
        raise MandateValidationError(f"Intent mandate expired at {intentMandate.validUntilTimestamp} (now: {now})")
    if executionMandate.settlementAmountPaise > intentMandate.maxBudgetPaise:
        raise MandateValidationError(f"Settlement amount exceeds max budget ({intentMandate.maxBudgetPaise})")
    if executionMandate.settlementAmountPaise > intentMandate.singleTransactionLimitPaise:
        raise MandateValidationError(f"Settlement amount exceeds single limit ({intentMandate.singleTransactionLimitPaise})")
    if cartMandate.totalPaise != executionMandate.settlementAmountPaise:
        raise MandateValidationError("Cart total does not match execution settlement amount")
    verifyMandateHashChain(intentMandate, cartMandate, executionMandate, raiseOnMismatch=True)


createSignedIntentMandate = createIntentMandate
createSignedCartMandate = createCartMandate
createSignedExecutionMandate = createExecutionMandate
createSignedAmendmentMandate = createAmendmentMandate


class AgentMandateBuilder:
    """AP2 Mandate lifecycle builder and validator factory."""

    createIntentMandate = staticmethod(createIntentMandate)
    createCartMandate = staticmethod(createCartMandate)
    createExecutionMandate = staticmethod(createExecutionMandate)
    createAmendmentMandate = staticmethod(createAmendmentMandate)
    computeHash = staticmethod(computeMandateHash)
    verifyHashChain = staticmethod(verifyMandateHashChain)
    validateInvariants = staticmethod(validateMandateInvariants)


__all__ = [
    "AgentMandateBuilder",
    "computeMandateHash",
    "createAmendmentMandate",
    "createCartMandate",
    "createExecutionMandate",
    "createIntentMandate",
    "createSignedAmendmentMandate",
    "createSignedCartMandate",
    "createSignedExecutionMandate",
    "createSignedIntentMandate",
    "validateMandateInvariants",
    "verifyMandateHashChain",
]

File: packages/buyerSdkTs/src/agentMandateBuilder.ts
import { v4 as uuidv4 } from "uuid";
import {
  defaultCurrency,
  defaultIntentValiditySeconds,
  mandateAmendPrefix,
  mandateCartPrefix,
  mandateExecPrefix,
  mandateIntentPrefix,
  millisPerSecond,
  signatureFieldKeys,
  uncategorizedCartItemCategory
} from "./sdkConstants.js";
import { AgentKeyManager } from "./agentKeyManager.js";
import { canonicalizeJson, computeSha256Digest } from "./jcsCanonicalizer.js";
import {
  ArithmeticDriftException,
  MandateVerificationError,
  type AmendmentMandate,
  type CartItem,
  type CartMandate,
  type ExecutionMandate,
  type IntentMandate,
  type TaxBreakdown
} from "./types.js";

export interface CreateIntentMandateParams {
  readonly mandateId?: string;
  readonly delegatedAgentDid: string;
  readonly maxBudgetPaise: number;
  readonly upiCircleDelegationToken: string;
  readonly singleTransactionLimitPaise: number;
  readonly authorizedCategories?: readonly string[];
  readonly validUntilTimestamp?: number;
  readonly nonce?: string;
  readonly timestamp?: number;
}

export interface CreateCartMandateParams {
  readonly cartId?: string;
  readonly merchantGstin: string;
  readonly merchantStateCode: string;
  readonly buyerDeliveryPincode: string;
  readonly buyerDeliveryStateCode: string;
  readonly items: readonly CartItem[];
  readonly taxableSubtotalPaise: number;
  readonly taxBreakdown: TaxBreakdown;
  readonly shippingPaise?: number;
  readonly discountPaise?: number;
  readonly totalPaise: number;
  readonly inventoryLockToken: string;
  readonly inventoryLockExpiresAt: number;
  readonly nonce?: string;
  readonly timestamp?: number;
}

export interface CreateExecutionMandateParams {
  readonly executionId?: string;
  readonly intentMandate: IntentMandate;
  readonly cartMandate: CartMandate;
  readonly settlementAmountPaise: number;
  readonly upiCircleToken: string;
  readonly nonce?: string;
  readonly timestamp?: number;
}

export interface CreateAmendmentMandateParams {
  readonly amendmentId?: string;
  readonly previousCartMandate: CartMandate;
  readonly newCartMandate: CartMandate;
  readonly substitutedSkuMapping: Readonly<Record<string, string>>;
  readonly priceDeltaPaise: number;
  readonly amendmentReason: string;
  readonly nonce?: string;
  readonly timestamp?: number;
}

export function computeMandateHash(mandate: Record<string, unknown>): string {
  const unsignedDict: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(mandate)) {
    if (!(signatureFieldKeys as readonly string[]).includes(key)) {
      unsignedDict[key] = value;
    }
  }
  const canonicalBytes = canonicalizeJson(unsignedDict);
  return computeSha256Digest(canonicalBytes);
}

export function createSignedIntentMandate(
  params: CreateIntentMandateParams,
  userSigner: AgentKeyManager
): IntentMandate {
  if (params.maxBudgetPaise <= 0 || params.singleTransactionLimitPaise <= 0) {
    throw new ArithmeticDriftException("Budget ceiling and transaction limit must be positive integer paise");
  }
  const ts = params.timestamp ?? Math.floor(Date.now() / millisPerSecond);
  const nonce = params.nonce ?? uuidv4().replace(/-/g, "");
  const mandateId = params.mandateId ?? `${mandateIntentPrefix}${uuidv4().replace(/-/g, "").slice(0, 16)}`;
  const validUntil = params.validUntilTimestamp ?? ts + defaultIntentValiditySeconds;
  const categories = params.authorizedCategories ?? [];

  const unsignedPayload = {
    authorizedCategories: categories,
    currency: defaultCurrency,
    delegatedAgentDid: params.delegatedAgentDid,
    mandateId,
    maxBudgetPaise: params.maxBudgetPaise,
    nonce,
    singleTransactionLimitPaise: params.singleTransactionLimitPaise,
    timestamp: ts,
    upiCircleDelegationToken: params.upiCircleDelegationToken,
    userDid: userSigner.getAgentDid(),
    validUntilTimestamp: validUntil
  };

  const userSignature = userSigner.signPayload(unsignedPayload);
  return { ...unsignedPayload, userSignature };
}

export const createIntentMandate = createSignedIntentMandate;

/**
 * Fills in the category the merchant is asserting. Python's CartItemSchema defaults this field
 * and always emits it, while JSON.stringify drops an undefined key entirely -- so leaving it out
 * here would make the same cart canonicalize to different bytes on the two sides and every
 * cross-SDK signature check would fail. Normalising once, before signing, keeps the wire format
 * identical whether or not the caller classified the SKU.
 */
function _withSignedCategory(item: CartItem): CartItem {
  return { ...item, category: item.category ?? uncategorizedCartItemCategory };
}

export function createSignedCartMandate(
  params: CreateCartMandateParams,
  merchantSigner: AgentKeyManager
): CartMandate {
  if (params.items.length === 0) {
    throw new Error("Cart must contain at least one item");
  }
  const ts = params.timestamp ?? Math.floor(Date.now() / millisPerSecond);
  const nonce = params.nonce ?? uuidv4().replace(/-/g, "");
  const cartId = params.cartId ?? `${mandateCartPrefix}${uuidv4().replace(/-/g, "").slice(0, 16)}`;

  const unsignedPayload = {
    buyerDeliveryPincode: params.buyerDeliveryPincode,
    buyerDeliveryStateCode: params.buyerDeliveryStateCode,
    cartId,
    discountPaise: params.discountPaise ?? 0,
    inventoryLockExpiresAt: params.inventoryLockExpiresAt,
    inventoryLockToken: params.inventoryLockToken,
    items: params.items.map(_withSignedCategory),
    merchantDid: merchantSigner.getAgentDid(),
    merchantGstin: params.merchantGstin,
    merchantStateCode: params.merchantStateCode,
    nonce,
    shippingPaise: params.shippingPaise ?? 0,
    taxBreakdown: params.taxBreakdown,
    taxableSubtotalPaise: params.taxableSubtotalPaise,
    timestamp: ts,
    totalPaise: params.totalPaise
  };

  const merchantSignature = merchantSigner.signPayload(unsignedPayload);
  return { ...unsignedPayload, merchantSignature };
}

export const createCartMandate = createSignedCartMandate;

/** The Execution Mandate exactly as it is signed: `model_dump()` minus `agentSignature`. */
export type UnsignedExecutionPayload = Omit<ExecutionMandate, "agentSignature">;

/**
 * Builds the nine-key payload an Execution Mandate signature covers, without signing it.
 *
 * Split out of `createSignedExecutionMandate` so that a caller who cannot sign in-process --
 * an MCP server handing exact bytes to an external agent that holds its own key -- derives the
 * payload from this one definition rather than restating it. The canonical shape already
 * exists in mandateFactory.py and buyerSdkPy/mandateModels.py; a fourth hand-rolled copy would
 * sign cleanly and fail verification at settlement with no useful error.
 *
 * `buyerAgentDid` is a parameter rather than a signer lookup precisely because the agent-held
 * path has a DID but no key.
 */
export function buildUnsignedExecutionPayload(
  params: CreateExecutionMandateParams,
  buyerAgentDid: string
): UnsignedExecutionPayload {
  const ts = params.timestamp ?? Math.floor(Date.now() / millisPerSecond);
  const nonce = params.nonce ?? uuidv4().replace(/-/g, "");
  const executionId = params.executionId ?? `${mandateExecPrefix}${uuidv4().replace(/-/g, "").slice(0, 16)}`;

  return {
    buyerAgentDid,
    cartMandateHash: computeMandateHash(params.cartMandate as unknown as Record<string, unknown>),
    currency: defaultCurrency,
    executionId,
    intentMandateHash: computeMandateHash(params.intentMandate as unknown as Record<string, unknown>),
    nonce,
    settlementAmountPaise: params.settlementAmountPaise,
    timestamp: ts,
    upiCircleToken: params.upiCircleToken
  };
}

export function createSignedExecutionMandate(
  params: CreateExecutionMandateParams,
  buyerAgentSigner: AgentKeyManager
): ExecutionMandate {
  const unsignedPayload = buildUnsignedExecutionPayload(params, buyerAgentSigner.getAgentDid());
  const agentSignature = buyerAgentSigner.signPayload(unsignedPayload);
  return { ...unsignedPayload, agentSignature };
}

export const createExecutionMandate = createSignedExecutionMandate;

export function createSignedAmendmentMandate(
  params: CreateAmendmentMandateParams,
  buyerAgentSigner: AgentKeyManager,
  merchantSigner: AgentKeyManager
): AmendmentMandate {
  const ts = params.timestamp ?? Math.floor(Date.now() / millisPerSecond);
  const nonce = params.nonce ?? uuidv4().replace(/-/g, "");
  const amendmentId = params.amendmentId ?? `${mandateAmendPrefix}${uuidv4().replace(/-/g, "").slice(0, 16)}`;

  const previousCartMandateHash = computeMandateHash(params.previousCartMandate as unknown as Record<string, unknown>);
  const newCartMandateHash = computeMandateHash(params.newCartMandate as unknown as Record<string, unknown>);

  const unsignedPayload = {
    amendmentId,
    amendmentReason: params.amendmentReason,
    newCartMandateHash,
    nonce,
    previousCartMandateHash,
    priceDeltaPaise: params.priceDeltaPaise,
    substitutedSkuMapping: params.substitutedSkuMapping,
    timestamp: ts
  };

  const canonicalBytes = canonicalizeJson(unsignedPayload);
  const agentSignature = buyerAgentSigner.signCanonicalBytes(canonicalBytes);
  const merchantSignature = merchantSigner.signCanonicalBytes(canonicalBytes);

  return { ...unsignedPayload, agentSignature, merchantSignature };
}

export const createAmendmentMandate = createSignedAmendmentMandate;

// Returns true when the chain holds. What happens when it does not is the caller's choice, and
// that choice used to be unavailable: this returned `boolean` but only ever returned `true` or
// threw, so `if (!verifyMandateChain(...))` typechecked and could never run its else branch. The
// parameter mirrors the Python SDK's `verifyMandateHashChain(..., raiseOnMismatch=True)`, and
// defaults to throwing so existing callers are unaffected.
export function verifyMandateChain(
  intentMandate: IntentMandate,
  cartMandate: CartMandate,
  executionMandate: ExecutionMandate,
  raiseOnMismatch: boolean = true
): boolean {
  try {
    _verifyIntentSignature(intentMandate, executionMandate);
    _verifyCartSignature(cartMandate, executionMandate);
    _verifyChainLinkage(intentMandate, cartMandate, executionMandate);
    return true;
  } catch (error) {
    // Only a failed verification becomes `false`. Anything else -- a malformed mandate that makes
    // canonicalization itself throw -- is a different problem and must not be reported as a
    // cleanly rejected chain.
    if (raiseOnMismatch || !(error instanceof MandateVerificationError)) {
      throw error;
    }
    return false;
  }
}

function _verifyIntentSignature(intentMandate: IntentMandate, executionMandate: ExecutionMandate): void {
  const computedIntentHash = computeMandateHash(intentMandate as unknown as Record<string, unknown>);
  if (computedIntentHash !== executionMandate.intentMandateHash) {
    throw new MandateVerificationError(
      `Intent mandate hash mismatch: expected ${executionMandate.intentMandateHash} (recorded in the execution mandate), got ${computedIntentHash} (recomputed now)`
    );
  }
}

function _verifyCartSignature(cartMandate: CartMandate, executionMandate: ExecutionMandate): void {
  const computedCartHash = computeMandateHash(cartMandate as unknown as Record<string, unknown>);
  if (computedCartHash !== executionMandate.cartMandateHash) {
    throw new MandateVerificationError(
      `Cart mandate hash mismatch: expected ${executionMandate.cartMandateHash} (recorded in the execution mandate), got ${computedCartHash} (recomputed now)`
    );
  }
}

function _verifyChainLinkage(
  intentMandate: IntentMandate,
  cartMandate: CartMandate,
  executionMandate: ExecutionMandate
): void {
  if (executionMandate.settlementAmountPaise !== cartMandate.totalPaise) {
    throw new MandateVerificationError(
      `Settlement amount ${executionMandate.settlementAmountPaise} does not match cart total ${cartMandate.totalPaise}`
    );
  }

  if (cartMandate.totalPaise > intentMandate.maxBudgetPaise) {
    throw new MandateVerificationError(
      `Cart total ${cartMandate.totalPaise} exceeds intent max budget ${intentMandate.maxBudgetPaise}`
    );
  }

  if (cartMandate.totalPaise > intentMandate.singleTransactionLimitPaise) {
    throw new MandateVerificationError(
      `Cart total ${cartMandate.totalPaise} exceeds single transaction limit ${intentMandate.singleTransactionLimitPaise}`
    );
  }

  if (executionMandate.timestamp > intentMandate.validUntilTimestamp) {
    throw new MandateVerificationError(
      `Intent mandate expired at ${intentMandate.validUntilTimestamp} (current: ${executionMandate.timestamp})`
    );
  }
}
